ios-base/Navigators/Navigator.swift

```diff
@@ -0,0 +1,204 @@
+  var currentViewController: UIViewController? { get set }
+
+  /// Convencience init to set your application starting screen.
+  init(with route: Route)
+
+  /**
+   Navigate from your current screen to a new route.
+   - Parameters:
+   - route: The destination route of your navigation action.
+   - transition: The transition type that you want to use.
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func navigate(to route: Route, with transition: TransitionType, animated: Bool, completion: (() -> Void)?)
+
+  /**
+   Navigate from your current screen to a new entire navigator. Can only push a router as a modal. Afterwards, other controllers can be pushed inside the presented Navigator.
+   - Parameters:
+   - Navigator: The destination navigator that you want to navigate to
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func navigate(to router: Navigator, animated: Bool, completion: (() -> Void)?)
+
+  /**
+   Handles backwards navigation through the stack.
+   */
+  func pop(animated: Bool)
+
+  /**
+   Handles backwards navigation through the stack.
+   - Parameters:
+   - route: The index of the route of your navigation action.
+   - animated: Animate the transition or not.
+   */
+  func popTo(index: Int, animated: Bool)
+
+  /**
+   Handles backwards navigation through the stack.
+   - Parameters:
+   - route: The destination route of your navigation action.
+   - animated: Animate the transition or not.
+   */
+  func popTo(route: Route, animated: Bool)
+
+  /**
+   Dismiss your current ViewController.
+   - Parameters:
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func dismiss(animated: Bool, completion: (() -> Void)?)
+}
+
+public extension Navigator {
+  func navigate(to route: Route, with transition: TransitionType, animated: Bool = true, completion: (() -> Void)? = nil) {
+    let viewController = route.screen
+    switch transition {
+    case .modal:
+      currentViewController?.present(viewController, animated: animated, completion: completion)
+      currentViewController = viewController
+    case .push:
+      rootViewController?.pushViewController(viewController, animated: animated)
+      currentViewController = viewController
+    case .reset:
+      rootViewController?.setViewControllers([viewController], animated: animated)
+      currentViewController = viewController
+    case .changeRoot:
+      UIView.animate(withDuration: 0.3) { [weak self] in
+        if let navigation = (viewController as? UINavigationController) {
+          UIApplication.shared.keyWindow?.rootViewController = navigation
+          self?.rootViewController = navigation
+          self?.currentViewController = self?.rootViewController?.topViewController
+        } else {
+          let navigationController = UINavigationController(rootViewController: viewController)
```

Comment:
Maybe we can move this few lines into a `embedIntoNavigationController` method since it is repeated.